src/main/java/org/truffleruby/core/numeric/IntegerNodes.java

```diff
@@ -1776,6 +1781,41 @@ protected int getLimit() {
 
     }
 
+    @Primitive(name = "mod_pow")
+    @NodeChild(value = "base", type = RubyNode.class)
+    @NodeChild(value = "exponent", type = RubyNode.class)
+    @NodeChild(value = "modulo", type = RubyNode.class)
+    public abstract static class ModPowNode extends PrimitiveNode {
+        @Child private FixnumOrBignumNode fixnumOrBignum = new FixnumOrBignumNode();
+
+        @CreateCast("base")
+        protected RubyNode baseToBigInteger(RubyNode base) {
+            return BigIntegerCastNode.create(base);
+        }
+
+        @CreateCast("exponent")
+        protected RubyNode exponentToBigInteger(RubyNode exponent) {
+            return BigIntegerCastNode.create(exponent);
+        }
+
+        @CreateCast("modulo")
+        protected RubyNode moduloToBigInteger(RubyNode modulo) {
+            return BigIntegerCastNode.create(modulo);
+        }
+
+        @Specialization(guards = "modulo.signum() < 0")
+        protected Object mod_pow_neg(BigInteger base, BigInteger exponent, BigInteger modulo) {
+            BigInteger result = base.modPow(exponent, modulo.negate());
+            return fixnumOrBignum.fixnumOrBignum(result.signum() == 1 ? result.add(modulo) : result);
+        }
+
+        @Specialization
```

Comment:
This should have a guard that it's positive.
Otherwise, this Specialization could be instantiated/marked as active, and then the next execution might use it even with a negative modulo.

Reply:
Also, it would be better and more efficient to handle 0 explicitly as a separate specialization, rather than relying on the automatic ArithmeticException translation.